dot/parachain/types/bitvec.go

```diff
@@ -1,132 +1,276 @@
+func (bv *BitVec) PushBits(bits []bool) {
+	// Pre-allocate space if needed
+	requiredBytes := (bv.len + len(bits) + 7) / 8
+	if requiredBytes > len(bv.bits) {
+		bytesToAdd := make([]byte, requiredBytes-len(bv.bits))
+		bv.bits = append(bv.bits, bytesToAdd...)
+	}
+
+	// Add each bit
+	for _, bit := range bits {
+		byteIndex := bv.len / 8
+		bitIndex := bv.len % 8
+
+		if bit {
+			bv.bits[byteIndex] |= 1 << bitIndex
 		}
+		bv.len++
+	}
+}
+
+// SetBit sets a bit at the specified index
+func (bv *BitVec) SetBit(index uint32, bit bool) error {
+	if index >= uint32(bv.len) {
+		return errors.New("index out of bounds")
 	}
 
-	return bytes
+	byteIndex := index / 8
+	bitIndex := index % 8
+
+	if bit {
+		bv.bits[byteIndex] |= 1 << bitIndex
+	} else {
+		bv.bits[byteIndex] &^= 1 << bitIndex
+	}
+	return nil
 }
 
-// bytesToBits converts a slice of bytes to a slice of bits
-func (bv *BitVec) setBits(b []byte, size uint) {
-	var bits []bool
-	for _, uint8val := range b {
-		end := size
-		if end > byteSize {
-			end = byteSize
-		}
-		size -= end
+// GetBit returns the bit at the specified index
+func (bv *BitVec) GetBit(index uint32) (bool, error) {
+	if index >= uint32(bv.len) {
+		return false, errors.New("index out of bounds")
+	}
 
-		for j := uint(0); j < end; j++ {
-			bit := (uint8val>>j)&1 == 1
-			bits = append(bits, bit)
+	byteIndex := index / 8
+	bitIndex := index % 8
+
+	return (bv.bits[byteIndex] & (1 << bitIndex)) != 0, nil
+}
+
+// ExtendByByte adds a byte to the BitVec. The bits are added in LSB0 order,
+func (bv *BitVec) ExtendByByte(b byte) {
+	// Pre-allocate space if needed
+	requiredBytes := (bv.len + 8 + 7) / 8
+	if requiredBytes > len(bv.bits) {
+		bytesToAdd := make([]byte, requiredBytes-len(bv.bits))
+		bv.bits = append(bv.bits, bytesToAdd...)
+	}
+
+	for i := 0; i < 8; i++ {
+		byteIndex := bv.len / 8
+		bitIndex := bv.len % 8
+
+		if (b & (1 << i)) != 0 {
+			bv.bits[byteIndex] |= 1 << bitIndex
 		}
+		bv.len++
 	}
-	bv.bits = bits
 }
 
-// MarshalSCALE fulfils the SCALE interface for encoding
+// MarshalSCALE encodes the BitVec into a byte slice
 func (bv BitVec) MarshalSCALE() ([]byte, error) {
-	buf := bytes.NewBuffer(nil)
-	encoder := scale.NewEncoder(buf)
-	if len(bv.bits) > bitVecMaxLength {
-		return nil, errBitVecTooLong
-	}
-	size := uint(len(bv.bits))
-	err := encoder.Encode(size)
-	if err != nil {
-		return nil, err
-	}
+	length := uint32(bv.len) // Get the current length of bits
+	header := encodeCompactLength(length)
 
-	bytes := bv.bytes()
-	_, err = buf.Write(bytes)
-	if err != nil {
-		return nil, err
+	// Combine the header and the actual bits into the result
+	result := make([]byte, len(header)+len(bv.bits))
+	copy(result, header)                // Copy the header into the result
+	copy(result[len(header):], bv.bits) // Copy the bits into the result
+
+	return result, nil
+}
+
+// encodeCompactLength encodes the length into a compact byte slice
+func encodeCompactLength(length uint32) []byte {
```

Comment:
I believe the current scale encoding implementation support encoding uint as compact encoding, so you don't need to implement another compact encoding 
```
var length uint = 10
scale.Marshal(length...)
```